Spigot-Server-Patches/0406-Implement-alternative-item-despawn-rate.patch

```diff
@@ -0,0 +1,106 @@
+     }
++
++    public boolean altItemDespawnRateEnabled;
++    public Map<Material, Integer> altItemDespawnRateMap;
++    private void altItemDespawnRate() {
++        String path = "alt-item-despawn-rate";
++
++            altItemDespawnRateEnabled = getBoolean(path + ".enabled", false);
++
++            Map<Material, Integer> altItemDespawnRateMapDefault = new EnumMap<>(Material.class);
++        altItemDespawnRateMapDefault.put(Material.COBBLESTONE, 300);
++        for(Material key : altItemDespawnRateMapDefault.keySet()) {
++            config.addDefault("world-settings.default." + path + ".items." + key, altItemDespawnRateMapDefault.get(key));
++        }
++
++            ConfigurationSection mapSection = config.getConfigurationSection("world-settings." + worldName + "." + path + ".items");
++        if(mapSection == null) {
++            mapSection = config.getConfigurationSection("world-settings.default." + path + ".items");
++        }
++        Map<String, Integer> rawMap = new HashMap<>();
++        for(String key : mapSection.getKeys(false)) {
++            int val = mapSection.getInt(key);
++            rawMap.put(key, val);
++        }
++
++            altItemDespawnRateMap = new EnumMap<>(Material.class);
++        for(String key : rawMap.keySet()) {
++            try {
++                altItemDespawnRateMap.put(Material.valueOf(key), rawMap.get(key));
++            } catch (Exception e) {
++                logError("Could not add item " + key + " to altItemDespawnRateItems: " + e.getMessage());
++            }
++        }
++        if(altItemDespawnRateEnabled) {
++            for(Material key : altItemDespawnRateMap.keySet()) {
++                log("alt-item-despawn-rate " + key + ": " + altItemDespawnRateMap.get(key));
++            }
++        }
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
+index 209169895..59c6586b2 100644
+--- a/src/main/java/net/minecraft/server/EntityItem.java
++++ b/src/main/java/net/minecraft/server/EntityItem.java
+@@ -5,6 +5,7 @@ import java.util.List;
+ import java.util.UUID;
+ import javax.annotation.Nullable;
+ // CraftBukkit start
++import org.bukkit.Material; // Paper
+ import org.bukkit.event.entity.EntityPickupItemEvent;
+ import org.bukkit.event.player.PlayerPickupItemEvent;
+ // CraftBukkit end
+@@ -127,6 +128,19 @@ public class EntityItem extends Entity {
+                 }
+             }
+ 
++            // Paper start - alternative item despawn rate
++            Material material = this.getItemStack().getBukkitStack().getType();
++            if (world.paperConfig.altItemDespawnRateEnabled && world.paperConfig.altItemDespawnRateMap.containsKey(material)) {
++                if(!this.world.isClientSide && this.age >= world.paperConfig.altItemDespawnRateMap.get(material)) {
++                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callItemDespawnEvent(this).isCancelled()) {
```

Comment:
avoid the duplication of death logic

(i.e use the spigot value if we don't have a config'd one)